appinventor/appengine/src/com/google/appinventor/server/storage/remote/RemoteStorage.java

```diff
@@ -0,0 +1,116 @@
+ */
+public abstract class RemoteStorage {
+  // We partition the bucket into different usage types, so specific lifecycle rules could
+  //   be applied by prefix, if required.
+  private final static String BUILD_OUTPUT_PREFIX = "build";
+  private final static String PROJECT_EXPORT_PREFIX = "export";
+
+  /**
+   * Generates a, usually, presigned URL to upload the file to the remote
+   *   storage server.
+   *
+   * @param objectKey object key in the remote storage server
+   * @return (presigned) upload URL
+   */
+  public abstract String generateUploadUrl(final String objectKey);
+
+  /**
+   * Generates an external URL to access the object. Depending on the
+   *   implementation, this may behave as a presigned URL or a "normal"
+   *   URL (if using a CDN, for example).
+   *
+   * @param objectKey object key in the remote storage server
+   * @return (presigned) get/retrieval URL
+   */
+  public abstract String generateRetrieveUrl(final String objectKey);
+
+  /**
+   * Generates a constant object key for a given specific project build output.
+   *
+   * @param target the type of target (Android)
+   * @param userId the user ID owning the project
+   * @param projectId the given project ID
+   * @param projectName the name of the project to store
+   * @param extensionName the extension name the build is on
+   * @return build/userId/projectId/target/projectName.extensionName
+   */
+  public final String getBuildOutputObjectKey(final String target, final String userId, final Long projectId,
+      final String projectName, final String extensionName) {
+    final String fileName = projectName + "." + extensionName;
+    return getBuildOutputObjectKey(target, userId, projectId, fileName);
+  }
+
+  /**
+   * Generates a constant object key for a given specific project build output.
+   *
+   * @param target the type of target (Android)
+   * @param userId the user ID owning the project
+   * @param projectId the given project ID
+   * @param fileName the file name to download
+   * @return build/userId/projectId/target/projectName.extensionName
+   */
+  public final String getBuildOutputObjectKey(final String target, final String userId, final Long projectId,
+      final String fileName) {
+    final String filePath = userId + "/" + projectId + "/" + target;
+    return BUILD_OUTPUT_PREFIX + "/" + filePath + "/" + fileName;
+  }
+
+  /**
+   * Generates a constant object key for a given specific project export.
+   *
+   * @param downloadKind the type of download
+   * @param userId the user ID owning the project
+   * @param fileName the name of the file to store
+   * @return export/userId/downloadKind/timestampHash/file.name
+   */
+  public final String getProjectExportObjectKey(final String downloadKind, final String userId, final String fileName) {
+    final String timestamp = String.valueOf(Instant.now().getEpochSecond());
+    final String timestampHash = generateFieldsHash(userId, downloadKind, fileName, timestamp);
```

Comment:
I'm introducing this as part of the object key to ensure we don't overwrite previous exports (in case the user still has the previous URL).

We can either keep it as it is, or sync it with how it's done for the build output, and just overwrite it. The previous presigned URL will still be valid, but will "point" to the new object already.